pkg/agent/resource/manager.go

```diff
@@ -0,0 +1,232 @@
+		log.GlobalLoggerConfig.Level(resCfgLogLevel).Apply()
+	}
+
+	if resCfgAdditionalTags != "" && !strings.EqualFold(cfg.TagsToPublish, resCfgAdditionalTags) {
+		cfg.TagsToPublish = resCfgAdditionalTags
+	}
+
+	// If config team is blank, check resource team name.  If resource team name is not blank, use resource team name
+	if resCfgTeamName != "" && !strings.EqualFold(cfg.TeamName, resCfgTeamName) {
+		cfg.TeamName = resCfgTeamName
+	}
+}
+
+func (a *agentResourceManager) onResourceChange() {
+	isChanged := (a.prevAgentResHash != 0)
+	agentResHash, _ := util.ComputeHash(a.agentResource)
+	if a.prevAgentResHash != 0 {
+		if a.prevAgentResHash == agentResHash {
+			isChanged = false
+		}
+	}
+	a.prevAgentResHash = agentResHash
+	if isChanged {
+		// merge agent resource config with central config
+		a.mergeResourceWithConfig()
+		if a.agentResourceChangeHandler != nil {
+			a.agentResourceChangeHandler()
+		}
+	}
+}
+
+func (a *agentResourceManager) getAgentResourceType() string {
+	agentType, ok := AgentTypesMap[a.cfg.GetAgentType()]
+	if ok {
+		return agentType
+	}
+	return ""
+}
+
+// GetAgentResource - returns the agent resource
+func (a *agentResourceManager) getAgentResource() (*v1.ResourceInstance, error) {
+	agentResourceType := a.getAgentResourceType()
+	agentResourceURL := a.cfg.GetEnvironmentURL() + "/" + agentResourceType + "/" + a.cfg.GetAgentName()
+
+	response, err := a.apicClient.ExecuteAPI(api.GET, agentResourceURL, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	agent := &v1.ResourceInstance{}
+	err = json.Unmarshal(response, agent)
+	if err != nil {
+		return nil, err
+	}
+	return agent, nil
+}
+
+func (a *agentResourceManager) updateAgentStatusAPI(resource interface{}, agentResourceType string) error {
+	buffer, err := json.Marshal(resource)
+	if err != nil {
+		return nil
+	}
+
+	subResURL := a.cfg.GetEnvironmentURL() + "/" + agentResourceType + "/" + a.cfg.GetAgentName() + "/status"
+	_, err = a.apicClient.ExecuteAPI(api.PUT, subResURL, nil, buffer)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (a *agentResourceManager) createAgentStatusSubResource(agentResourceType, status, prevStatus, message string) *v1.ResourceInstance {
+	switch agentResourceType {
+	case v1alpha1.DiscoveryAgentResourceName:
+		agentRes := createDiscoveryAgentStatusResource(a.cfg.GetAgentName(), status, prevStatus, message)
+		resourceInstance, _ := agentRes.AsInstance()
+		return resourceInstance
+	case v1alpha1.TraceabilityAgentResourceName:
+		agentRes := createTraceabilityAgentStatusResource(a.cfg.GetAgentName(), status, prevStatus, message)
+		resourceInstance, _ := agentRes.AsInstance()
+		return resourceInstance
+	case v1alpha1.GovernanceAgentResourceName:
+		agentRes := createGovernanceAgentStatusResource(a.cfg.GetAgentName(), status, prevStatus, message)
+		resourceInstance, _ := agentRes.AsInstance()
+		return resourceInstance
+	default:
+		panic(ErrUnsupportedAgentType)
```

Comment:
Can we return an error instead of a panic?

Reply:
Done